torchbenchmark/util/framework/vision/model_factory.py

```diff
@@ -71,8 +71,7 @@ def get_module(self):
         return self.model, self.example_inputs
 
     def forward(self):
-        with torch.no_grad():
-            self.example_outputs = (torch.rand_like(self.model(*self.example_inputs)), )
+        self.example_outputs = (torch.rand_like(self.model(*self.example_inputs)), )
```

Comment:
I am wondering why we are removing the `torch.no_grad()` context here? They are added by https://github.com/pytorch/benchmark/pull/1558

Reply:
vision forward will be called in `_invoke_staged_train_test`, since vision models train is staged train https://github.com/pytorch/benchmark/blob/main/torchbenchmark/util/model.py#L299. It will call `backward` then. However, `torch.no_grad` will make the input no grad, so `backward` will run into failure.

Here I have a concern as well, why there is no failure of `amp_fp16` in cuda device when running vision training test.

Reply:
I am thinking here we are generating `example_output`, not `example_input`. The `example_input` tensor is still with grad. Is my understanding correct?

Reply:
Hi @xuzhao9 , thanks for the comments.
```
    def forward(self):
        with torch.no_grad():
            self.example_outputs = (torch.rand_like(self.model(*self.example_inputs)), )
        for data, target in zip(self.example_inputs, self.example_outputs):
            pred = self.model(data)
            return self.loss_fn(pred, target)
```

This is a confused issue. With torch.no_grad(), when running `python run.py alexnet -d cpu -t train --precision amp_bf16`, it will encounter the failure as below:
```
Traceback (most recent call last):
  File "run.py", line 623, in <module>
    main()  # pragma: no cover
  File "run.py", line 593, in main
    run_one_step(
  File "run.py", line 173, in run_one_step
    func()
  File "/workspace/benchmark/torchbenchmark/util/model.py", line 309, in invoke
    return self._invoke_staged_train_test(num_batch=self.num_batch)
  File "/workspace/benchmark/torchbenchmark/util/model.py", line 301, in _invoke_staged_train_test
    self.backward(losses)
  File "/workspace/benchmark/torchbenchmark/util/framework/vision/model_factory.py", line 81, in backward
    loss.backward()
  File "/workspace/pytorch/torch/_tensor.py", line 525, in backward
    torch.autograd.backward(
  File "/workspace/pytorch/torch/autograd/__init__.py", line 267, in backward
    _engine_run_backward(
  File "/workspace/pytorch/torch/autograd/graph.py", line 744, in _engine_run_backward
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn
```
I print out `pred.requires_grad` and `target.requires_grad`, they are all `False`. Then in backward, it will fail. It's weird that in a simple UT of the forward logic, `pred.requires_grad` is always True. But in torchbench, `pred.requires_grad` will be False, it has to remove `with torch.no_grad()`.
```
import torch
model = torch.nn.Linear(2, 2)
input = torch.randn(2, 2)
inputs = [input]

with torch.no_grad():
    outputs = (torch.rand_like(model(*inputs)), )
for data, target in zip(inputs, outputs):
    pred = model(data)

print(data.requires_grad)
print(pred.requires_grad)
print(target.requires_grad)
```
I'm not sure how to fix this in torchbench. Could you please help on this?

Reply:
Can I have the command to reproduce this failure? I tested in https://github.com/pytorch/benchmark/pull/2206, and it actually runs without throwing the backward/gradient exception.

Reply:
@xuzhao9 `python run.py alexnet -d cpu -t train --precision amp_bf16` can reproduce the failure.

Reply:
@yanbing-j Yeah strangely this only happens on alexnet, not `resnet18` or other torchvision models. I am looking into it, perhaps it is a torchvision bug.

Reply:
@yanbing-j I am thinking if this is only `alexnet` issue, we could just keep `with torch.no_grad()` and deal with this model later.

Reply:
@xuzhao9 Thanks for the investigation. Could you please give some suggestions of dealing with `alexnet` later?

Reply:
@yanbing-j I think I have found the issue. For alexnet, the `pred = self.model(data)` call returns a tensor without grad even if `data` requires grad. Changing it to `pred = self.model(data).requires_grad_(True)` will fix this issue. Can you help try it?

Reply:
Here is an example of my change that passes the test: https://github.com/pytorch/benchmark/pull/2206/files

Reply:
Thanks! The PR looks good to me now